ocaml/xapi/xapi_pgpu_helpers.ml

```diff
@@ -181,24 +186,38 @@ let assert_destination_pgpu_is_compatible_with_vm ~__context ~vm ~host ?remote (
       if supported then pgpu
       else get_first_pgpu_of_implementation vgpu implementation rest
   in
-  let vgpus = Db.VM.get_VGPUs ~__context ~self:vm in
-  List.iter (fun vgpu ->
+  let test_nvidia_compatibility vgpu pgpu = 
+    let pgpu_metadata =
+      let key = Xapi_gpumon.Nvidia.key in
+      try
+        get_compatibility_metadata pgpu
+        |> List.assoc key
+      with Not_found -> 
+        debug "Key %s is missing from the compatibility_metadata for pgpu %s on the host %s." key (Ref.string_of pgpu) (Ref.string_of host);
+        raise Api_errors.(Server_error (nvidia_tools_error, [Ref.string_of host]))
+    in
+    Xapi_gpumon.Nvidia.assert_pgpu_is_compatible_with_vm ~__context ~vm ~vgpu ~dest_host:host ~encoded_pgpu_metadata:pgpu_metadata
+  in
+  match vgpu_map with
+  | [] ->
+    let vgpus = Db.VM.get_VGPUs ~__context ~self:vm in
+    List.iter (fun vgpu ->
+      match vgpu_impl vgpu with
+      | `passthrough | `gvt_g | `mxgpu -> ();
+      | `nvidia -> 
+        let pgpu =
+          get_pgpus_of_host host
+          |> get_first_pgpu_of_implementation vgpu `nvidia
+        in test_nvidia_compatibility vgpu pgpu;
+      ) vgpus
+  | vgpus ->
+    List.iter (fun (vgpu, gpu_group) ->
       match vgpu_impl vgpu with
       | `passthrough | `gvt_g | `mxgpu -> ();
-      | `nvidia ->
-        let pgpu_metadata =
-          let key = Xapi_gpumon.Nvidia.key in
-          let pgpu =
-            get_pgpus host
-            |> get_first_pgpu_of_implementation vgpu `nvidia
-          in
-          try
-            get_compatibility_metadata pgpu
-            |> List.assoc key
-          with Not_found ->
-            debug "Key %s is missing from the compatibility_metadata for pgpu %s on the host %s." key (Ref.string_of pgpu) (Ref.string_of host);
-            raise Api_errors.(Server_error (nvidia_tools_error, [Ref.string_of host]))
-        in 
-        Xapi_gpumon.Nvidia.assert_pgpu_is_compatible_with_vm ~__context ~vm ~vgpu ~dest_host:host ~encoded_pgpu_metadata:pgpu_metadata
-    ) vgpus
+      | `nvidia -> 
+        let pgpu =
+          get_pgpus_in_gpu_group gpu_group
+          |> get_first_pgpu_of_implementation vgpu `nvidia
+        in test_nvidia_compatibility vgpu pgpu;
+      ) vgpus
```

Comment:
It looks to me as if some refactoring could be worthwhile here: the two uses of `List.iter` are almost identical, differing only in that one has `get_pgpus_of_host host` where the other has `get_pgpus_in_gpu_group gpu_group`.

Maybe something like
```
List.iter (fun (vgpu, candidate_gpus) ->
  ...
  | `nvidia -> let pgpu = candidate_gpus |> get_first_pgpu_of_implementation vgpu `nvidia
    ...
)
```
with `candidate_gpus` coming from either
`get_pgpus_of_host host` or ` get_pgpus_in_gpu_group gpu_group`
(or maybe use `get_candidate_gpus: unit -> gpu list`)